dotnet/src/Connectors/Connectors.Ollama/Services/OllamaChatCompletionService.cs

```diff
@@ -119,48 +381,403 @@ public async IAsyncEnumerable<StreamingChatMessageContent> GetStreamingChatMessa
+                properties.Add(parameter.Name, parameter.Schema ?? GetDefaultSchemaForTypelessParameter(parameter.Description));
+                if (parameter.IsRequired)
+                {
+                    required.Add(parameter.Name);
+                }
+            }
+
+            string serializedParametersSchema = JsonSerializer.Serialize(new
+            {
+                type = "object",
+                required,
+                properties,
+            });
+
+            resultParameters = JsonSerializer.Deserialize<Parameters>(serializedParametersSchema)!;
+        }
+
+        tool.Function.Parameters = resultParameters;
+
+        return tool;
+    }
+
+    /// <summary>Gets a <see cref="KernelJsonSchema"/> for a typeless parameter with the specified description, defaulting to typeof(string)</summary>
+    private static KernelJsonSchema GetDefaultSchemaForTypelessParameter(string? description)
+    {
+        // If there's a description, incorporate it.
+        if (!string.IsNullOrWhiteSpace(description))
+        {
+            return KernelJsonSchemaBuilder.Build(null, typeof(string), description);
+        }
+
+        // Otherwise, we can use a cached schema for a string with no description.
+        return s_stringNoDescriptionSchema;
+    }
+
+    private static List<Message> CreateChatCompletionMessages(OllamaPromptExecutionSettings executionSettings, ChatHistory chatHistory)
+    {
+        List<Message> ollamaMessages = [];
+
+        foreach (var chatMessageContent in chatHistory)
+        {
+            ollamaMessages.AddRange(CreateRequestMessages(chatMessageContent));
+        }
+
+        return ollamaMessages;
+    }
 
-    private static ChatRequest CreateChatRequest(ChatHistory chatHistory, OllamaPromptExecutionSettings settings, string selectedModel)
+    private static List<Message> CreateRequestMessages(ChatMessageContent message)
     {
-        var messages = new List<Message>();
-        foreach (var chatHistoryMessage in chatHistory)
+        // If the abstraction message preserves the original Ollama message, use it as is.
+        if (message.InnerContent is Message ollamaMessage)
+        {
+            return [ollamaMessage];
+        }
+
+        if (message.Role == AuthorRole.System)
+        {
+            return [new Message(ChatRole.System, message.Content ?? string.Empty)];
+        }
+
+        if (message.Role == AuthorRole.Tool)
         {
-            ChatRole role = ChatRole.User;
-            if (chatHistoryMessage.Role == AuthorRole.System)
+            // Handling function results represented by the TextContent type.
+            // Example: new ChatMessageContent(AuthorRole.Tool, content, metadata: new Dictionary<string, object?>(1) { { OpenAIChatMessageContent.ToolIdProperty, toolCall.Id } })
+            if (message.Metadata?.TryGetValue(OllamaChatCompletionService.ToolIdProperty, out object? toolId) is true &&
+                toolId?.ToString() is string toolIdString)
```

Comment:
Passing the function result from the function caller down to the connector via metadata collection is how it was originally done for the OpenAI connector. The connector still has it for backward compatibility. New connectors should not handle these cases and should use `FunctionResultContent` to represent the function invocation result.